Zend/zend_execute_API.c

```diff
@@ -1479,21 +1479,25 @@ void zend_unset_timeout(void) /* {{{ */
 }
 /* }}} */
 
-zend_class_entry *zend_fetch_class(zend_string *class_name, int fetch_type) /* {{{ */
+zend_class_entry *zend_fetch_class_with_scope(zend_string *class_name, int fetch_type, zend_class_entry *scope) /* {{{ */
 {
-	zend_class_entry *ce, *scope;
+	zend_class_entry *ce;
 	int fetch_sub_type = fetch_type & ZEND_FETCH_CLASS_MASK;
 
 check_fetch_type:
 	switch (fetch_sub_type) {
 		case ZEND_FETCH_CLASS_SELF:
-			scope = zend_get_executed_scope();
+			if (scope == NULL) {
```

Comment:
I think we could get into trouble here if the expression is supposed to be evaluated at global scope, in which case `scope` will be NULL and will fall back to the executed scope here. So we'd have to distinguish an explicit scope vs the fallback to the executed scope separately.

Reply:
Ah, forgot about global being NULL. Seems like the signature should allow providing the explicit vs execution scope? I'm out of my depth though so feel free to change/modify this.